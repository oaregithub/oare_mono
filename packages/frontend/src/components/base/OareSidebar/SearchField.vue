<template>
  <div class="d-flex flex-column">
    <div class="d-flex">
      <h4>{{ label }}</h4>
      <v-dialog v-model="helpOpen" width="500">
        <v-card>
          <v-card-title>Help</v-card-title>
          <v-card-text>
            <slot name="help-body"></slot>
          </v-card-text>
        </v-card>
      </v-dialog>
    </div>
    <v-text-field
      :value="value"
      @input="$emit('input', $event)"
      outlined
      filled
      :background-color="disabled ? 'grey' : 'white'"
      light
      dense
      :disabled="disabled"
    />
  </div>
</template>

<script lang="ts">
import { defineComponent, ref } from '@vue/composition-api';
export default defineComponent({
  props: {
    label: {
      type: String,
      required: true,
    },
    value: {
      type: String,
      required: true,
    },
    disabled: {
      type: Boolean,
      default: false,
    },
  },
  setup() {
    const helpOpen = ref(false);

    return {
      helpOpen,
    };
  },
});
</script>

<style></style>
