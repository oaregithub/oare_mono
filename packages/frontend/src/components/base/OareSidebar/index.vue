<template>
  <div class="white--text pa-3 d-flex flex-column">
    <h2 class="mb-2">Quick Queries</h2>
    <router-link to="/search/texts" class="white--text no-underline">
      Go to advanced queries <v-icon>mdi-icon-right</v-icon>
    </router-link>
    <search-field
      class="test-text-input"
      label="Text Name"
      v-model="text"
      @keyup.enter.native="performSearch"
      :disabled="textSearchDisabled"
    />
    <search-field
      class="mt-n3 test-transliteration-input"
      label="Transliteration"
      v-model="translit"
      @keyup.enter.native="performSearch"
      :disabled="textSearchDisabled"
      :displayHelp="true"
    />
    <search-field
      class="mt-n3 test-words-input"
      label="Words"
      v-model="word"
      @keyup.enter.native="performSearch"
      :disabled="dictionaryDisabled"
      v-if="isAdmin"
    />
    <v-btn color="error" class="mb-3 test-clear-btn" @click="clearSearch"
      >Clear</v-btn
    >
    <v-btn
      color="info"
      class="mb-3 test-search-btn"
      :disabled="!canSearch"
      @click="performSearch"
      >Search</v-btn
    >
  </div>
</template>

<script lang="ts">
import { defineComponent, computed, ref } from '@vue/composition-api';
import SearchField from './SearchField.vue';
import sl from '@/serviceLocator';

export default defineComponent({
  name: 'OareSidebar',
  components: {
    SearchField,
  },
  setup() {
    const store = sl.get('store');
    const text = ref('');
    const translit = ref('');
    const word = ref('');

    const textSearchDisabled = computed(() => word.value.trim() !== '');
    const dictionaryDisabled = computed(() => {
      return text.value.trim() !== '' || translit.value.trim() !== '';
    });

    const performSearch = async () => {
      const routeName = dictionaryDisabled.value
        ? '/search/texts'
        : '/search/dictionary';

      const { protocol, pathname, host } = window.location;
      const urlParams = new URLSearchParams();

      if (text.value) {
        urlParams.set('title', text.value);
      }

      if (translit.value) {
        urlParams.set('translit', translit.value);
      }

      if (word.value) {
        urlParams.set('dictionary', word.value);
      }

      const newUrl = `${protocol}//${host}${routeName}?${urlParams.toString()}`;
      window.location.href = newUrl;
    };

    const isAdmin = computed(() => store.getters.isAdmin);

    const canSearch = computed(
      () => text.value || translit.value || word.value
    );

    const clearSearch = () => {
      text.value = '';
      translit.value = '';
      word.value = '';
    };

    return {
      text,
      translit,
      word,
      performSearch,
      textSearchDisabled,
      dictionaryDisabled,
      clearSearch,
      canSearch,
      isAdmin,
    };
  },
});
</script>

<style scoped>
.no-underline {
  text-decoration: none;
}
</style>
