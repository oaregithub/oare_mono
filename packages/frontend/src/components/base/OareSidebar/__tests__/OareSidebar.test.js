import Vuetify from 'vuetify';
import VueCompositionApi from '@vue/composition-api';
import { mount, createLocalVue } from '@vue/test-utils';
import OareSidebar from '../index.vue';

const vuetify = new Vuetify();
const localVue = createLocalVue();
localVue.use(VueCompositionApi);

describe('OareSidebar test', () => {
  const createWrapper = (isAdmin = true) =>
    mount(OareSidebar, {
      localVue,
      vuetify,
      stubs: ['router-link'],
      propsData: {
        store: {
          getters: {
            isAdmin,
          },
        },
      },
    });

  it('matches snapshot', () => {
    expect(createWrapper()).toMatchSnapshot();
  });

  it('search button is disabled when inputs are all blank', () => {
    const searchBtn = createWrapper().find('.test-search-btn');
    expect(searchBtn.element).toHaveClass('v-btn--disabled');
  });

  it('Words search input is disabled when Text search field has value', async () => {
    const wrapper = createWrapper();
    const textInput = wrapper.find('.test-text-input').find('input');

    await textInput.setValue('CCT 1 1a');
    const wordsInput = wrapper.find('.test-words-input').find('input');
    expect(wordsInput.element).toBeDisabled();
  });

  it('Words search input is disabled when Transliteration search field has value', async () => {
    const wrapper = createWrapper();
    const transInput = wrapper
      .find('.test-transliteration-input')
      .find('input');

    await transInput.setValue('LUGAL');
    const wordsInput = wrapper.find('.test-words-input').find('input');
    expect(wordsInput.element).toBeDisabled();
  });

  it('Text and Transliteration fields are disabled when Words has value', async () => {
    const wrapper = createWrapper();
    const wordsInput = wrapper.find('.test-words-input').find('input');
    await wordsInput.setValue('ababum');

    const textInput = wrapper.find('.test-text-input').find('input');
    const transInput = wrapper
      .find('.test-transliteration-input')
      .find('input');
    expect(textInput.element).toBeDisabled();
    expect(transInput.element).toBeDisabled();
  });

  it('Clear button clears text and transliteration fields', async () => {
    const wrapper = createWrapper();
    const textInput = wrapper.find('.test-text-input').find('input');
    const transInput = wrapper
      .find('.test-transliteration-input')
      .find('input');

    await textInput.setValue('CCT');
    await transInput.setValue('LUGAL');

    const clearButton = wrapper.find('.test-clear-btn');
    await clearButton.trigger('click');

    expect(textInput.element).toHaveValue('');
    expect(transInput.element).toHaveValue('');
  });

  it('Clear button clears Words field', async () => {
    const wrapper = createWrapper();
    const wordsInput = wrapper.find('.test-words-input').find('input');
    await wordsInput.setValue('ababum');

    const clearButton = wrapper.find('.test-clear-btn');
    await clearButton.trigger('click');

    expect(wordsInput.element).toHaveValue('');
  });

  it('search button is disabled with no input', () => {
    const wrapper = createWrapper();
    const searchButton = wrapper.find('.test-search-btn');
    expect(searchButton.element).toHaveClass('v-btn--disabled');
  });

  it('search button is not disabled with Text input value', async () => {
    const wrapper = createWrapper();
    const textInput = wrapper.find('.test-text-input').find('input');
    await textInput.setValue('CCT');

    const searchButton = wrapper.find('.test-search-btn');
    expect(searchButton.element).not.toHaveClass('v-btn--disabled');
  });

  it('search button is not disabled with transliteration input value', async () => {
    const wrapper = createWrapper();
    const transInput = wrapper
      .find('.test-transliteration-input')
      .find('input');
    await transInput.setValue('LUGAL');

    const searchButton = wrapper.find('.test-search-btn');
    expect(searchButton.element).not.toHaveClass('v-btn--disabled');
  });

  it('search button is not disabled with words input value', async () => {
    const wrapper = createWrapper();
    const wordsInput = wrapper.find('.test-words-input input');
    await wordsInput.setValue('ababum');

    const searchButton = wrapper.find('.test-search-btn');
    expect(searchButton.element).not.toHaveClass('v-btn--disabled');
  });

  it('does not show words input if not an admin', () => {
    const wrapper = createWrapper(false);
    expect(wrapper.find('.test-words-input').exists()).toBe(false);
  });
});
