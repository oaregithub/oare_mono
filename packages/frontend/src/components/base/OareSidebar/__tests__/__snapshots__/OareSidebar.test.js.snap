// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`OareSidebar test matches snapshot 1`] = `
<div class="white--text pa-3 d-flex flex-column">
  <h2 class="mb-2">Quick Queries</h2>
  <router-link-stub to="/search/texts" class="white--text no-underline">
    Go to advanced queries <i aria-hidden="true" class="v-icon notranslate mdi mdi-icon-right theme--light"></i></router-link-stub>
  <div class="d-flex flex-column test-text-input">
    <div class="d-flex">
      <h4>Text</h4>
      <div role="dialog" class="v-dialog__container">
        <!---->
      </div>
    </div>
    <div class="v-input v-input--dense theme--light v-text-field v-text-field--single-line v-text-field--filled v-text-field--enclosed v-text-field--outlined">
      <div class="v-input__control">
        <div class="v-input__slot white">
          <fieldset aria-hidden="true">
            <legend><span>​</span></legend>
          </fieldset>
          <div class="v-text-field__slot"><input id="input-10" type="text"></div>
        </div>
        <div class="v-text-field__details">
          <div class="v-messages theme--light">
            <transition-group-stub tag="div" name="message-transition" class="v-messages__wrapper"></transition-group-stub>
          </div>
        </div>
      </div>
    </div>
  </div>
  <div class="d-flex flex-column mt-n3 test-transliteration-input">
    <div class="d-flex">
      <h4>Transliteration</h4>
      <div role="dialog" class="v-dialog__container">
        <!---->
      </div>
    </div>
    <div class="v-input v-input--dense theme--light v-text-field v-text-field--single-line v-text-field--filled v-text-field--enclosed v-text-field--outlined">
      <div class="v-input__control">
        <div class="v-input__slot white">
          <fieldset aria-hidden="true">
            <legend><span>​</span></legend>
          </fieldset>
          <div class="v-text-field__slot"><input id="input-15" type="text"></div>
        </div>
        <div class="v-text-field__details">
          <div class="v-messages theme--light">
            <transition-group-stub tag="div" name="message-transition" class="v-messages__wrapper"></transition-group-stub>
          </div>
        </div>
      </div>
    </div>
  </div>
  <div class="d-flex flex-column mt-n3 test-words-input">
    <div class="d-flex">
      <h4>Words</h4>
      <div role="dialog" class="v-dialog__container">
        <!---->
      </div>
    </div>
    <div class="v-input v-input--dense theme--light v-text-field v-text-field--single-line v-text-field--filled v-text-field--enclosed v-text-field--outlined">
      <div class="v-input__control">
        <div class="v-input__slot white">
          <fieldset aria-hidden="true">
            <legend><span>​</span></legend>
          </fieldset>
          <div class="v-text-field__slot"><input id="input-20" type="text"></div>
        </div>
        <div class="v-text-field__details">
          <div class="v-messages theme--light">
            <transition-group-stub tag="div" name="message-transition" class="v-messages__wrapper"></transition-group-stub>
          </div>
        </div>
      </div>
    </div>
  </div> <button type="button" class="mb-3 test-clear-btn v-btn v-btn--contained theme--light v-size--default error"><span class="v-btn__content">Clear</span></button> <button type="button" disabled="disabled" class="mb-3 test-search-btn v-btn v-btn--contained v-btn--disabled theme--light v-size--default"><span class="v-btn__content">Search</span></button>
</div>
`;
